### DISCUSSION TITLE
Can Factories capturing self cause circular references to occur on containers?

### DISCUSSION THREAD

**Discussion by rsaccone** [hmlongco/Factory#70] [Q&A] [answered]
I have a question about how best to refer to another factory from a factory.  Say I have something like this.

```
extension SharedContainer {
   var storage: Factory<Storage> {
           self { PersistentStorage() }
    }
  
   var recommendationsCoordinator: Factory<RecommendationsCoordinator> {
           self { RecommendationsCoordinator(storage: self.storage()) }
    }
}

// Somewhere in code
let recommendations = Container.shared.recommnedationsCoordinator()
```
    
Inside of the recommendationsCoordinator factory I am referencing storage via self instead of making any assumptions about it being in Container.shared. I did this  in case I do use a custom container down the line.  However what I worry about is the explicit use of self in "self.storage()" capturing a circular reference and things not cleaning up property in the case of a move to a custom container.  I realize that as long I use Container.shared this wouldn't be an issue.  Is the circular reference a worry and if so what is the best way to handle the issue?

Thanks,
Rob

**Comment by hmlongco** [comment, marked as answer]
It's not a worry. I may need to document this more, but as mentioned in the readme Factories are transient; created when needed, used, and then quickly discarded.

The definition of `recommendationsCoordinator` is a computed value. Hence its reference to self doesn't really exist until it's actually called. In a similar vein, that reference to the container in `{ RecommendationsCoordinator(storage: self.storage()) }` isn't captured until the Factory for `recommendationsCoordinator` is created and returned.

If that Factory is evaluated (Container.shared.recommnedationsCoordinator()), then the closure will be called, the parameter `self.storage()` will be evaluated,.and the process starts over, recursively, for the `storage` Factory.

That Factory returns its service and is discarded, its job complete. The `recommendationsCoordinator` creates and returns its instance, and that Factory is discarded in turn, its job complete. 

Everything's unwound, and the only thing now left is your reference to your coordinator in `recommendations`.

Just keep the temporal aspect in mind. The only thing you really shouldn't do it ask for a Factory and then stash it somewhere, as that Factory contains a hard reference to its container over its (usually brief) lifetime.

**Reply by rsaccone** [reply]
Thanks for the detailed response.  I overlooked the fact that the property is a computed value and I do remember your comment about the Factories as being transient.

**Reply by hmlongco** [reply]
Again, no worries. Keep in mind that questions like these also help me figure out what I need to emphasize or document better.

So much appreciated!
